chia/wallet/nft_wallet/nft_off_chain.py

```diff
@@ -0,0 +1,115 @@
+from chia.util.default_root import DEFAULT_ROOT_PATH
+from chia.util.hash import std_hash
+from chia.wallet.nft_wallet.nft_info import NFTCoinInfo
+from chia.wallet.nft_wallet.uncurry_nft import UncurriedNFT
+
+DEFAULT_PREFIX_HASH_LENGTH = 3
+DEFAULT_CACHE_PATH = "./nft_cache"
+CACHE_PATH_KEY = "nft_metadata_cache_path"
+PREFIX_HASH_LENGTH_KEY = "nft_metadata_cache_hash_length"
+
+
+log = logging.getLogger(__name__)
+
+
+async def fetch_off_chain_metadata(nft_coin_info: NFTCoinInfo, ignore_size_limit: bool) -> Optional[str]:
+    uncurried_nft: Optional[UncurriedNFT] = UncurriedNFT.uncurry(*nft_coin_info.full_puzzle.uncurry())
+    if uncurried_nft is None:
+        log.error(f"Cannot fetch off-chain metadata, {nft_coin_info.nft_id} is not a NFT.")
+        return None
+    for uri in uncurried_nft.meta_uris.as_python():  # pylint: disable=E1133
+        async with aiohttp.ClientSession() as session:
+            async with session.get(str(uri, "utf-8")) as response:
+                if response.status == 200:
+                    text = await response.text()
+                    if ignore_size_limit or sys.getsizeof(text) / 1024 / 1024 < 1:
+                        return text
+    return None
+
+
+def read_off_chain_metadata(nft_coin_info: NFTCoinInfo, config: Optional[Dict[str, Any]] = None) -> Optional[str]:
+    # Read metadata from disk cache
+    file_name = get_cached_filename(nft_coin_info.nft_id, config)
+    if not file_name.exists():
+        return None
+    try:
+        text = file_name.read_text()
+        if verify_metadata(text, nft_coin_info.full_puzzle):
+            return text
+        else:
+            return None
+    except Exception:
+        log.exception(f"Cannot load cached metadata of {nft_coin_info.nft_id.hex()}.")
+        return None
+
+
+def delete_off_chain_metadata(nft_id: bytes32, config: Optional[Dict[str, Any]] = None) -> None:
+    file_name = get_cached_filename(nft_id, config)
+    if file_name.exists():
+        file_name.unlink()
+        log.info(f"Deleted off-chain metadata of {nft_id.hex()}")
+
+
+def write_off_chain_metadata(nft_id: bytes32, metadata: str, config: Optional[Dict[str, Any]] = None) -> None:
+    file_name = get_cached_filename(nft_id, config)
+    file_name.write_text(metadata)
+
+
+async def get_off_chain_metadata(
+    nft_coin_info: NFTCoinInfo, config: Optional[Dict[str, Any]] = None, ignore_size_limit: bool = False
+) -> Optional[str]:
+    try:
+        # Check if the metadata is in disk cache
+        metadata = read_off_chain_metadata(nft_coin_info, config)
+        if metadata is not None:
+            return metadata
+        log.debug(f"{nft_coin_info.nft_id.hex()} is not in cache, downloading now ...")
+        metadata = await fetch_off_chain_metadata(nft_coin_info, ignore_size_limit)
+        if metadata is None:
+            log.error(f"Cannot find off-chain metadata of {nft_coin_info.nft_id.hex()}.")
+            return None
```

Comment:
I think if `fetch_off_chain_metadata` always returns None due to bad connection/not existing anymore, we still try to fetch it again forever whenever we call `get_nft_info_from_puzzle`, right?

Reply:
Yes, if there isn't a valid cached file it will keep fetching until success.